File: day07/main.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <numeric>
#include <sstream>
#include  <regex>
#include <optional>
#include <functional>
#include <set>
#include <unordered_map>
#include <map>

void part1(std::vector<std::pair<std::string, int64_t>> hands) {

    auto hand_type = [](const std::string& hand) {
        std::map<char, int> counts;
        for(char c: hand) ++counts[c];
        std::vector<int> type;
        for(const auto& elem: counts)
            type.push_back(elem.second);
        std::sort(type.begin(), type.end(), std::greater<int>());
        return type;
    };

    auto hand_level = [](const std::string& hand) {
        static const std::string order = "23456789TJQKA";
        std::vector<int> result;
        for(char c: hand) {
            auto pos = order.find(c);
            assert(pos!=std::string::npos);
            result.push_back(pos);
        }
        return result;
    };


    std::sort(hands.begin(), hands.end(), [&hand_type, &hand_level](const std::pair<std::string, int64_t>& a, const std::pair<std::string, int64_t>& b) {

        auto a_type = hand_type(a.first);
        auto b_type = hand_type(b.first);

        if(a_type==b_type) {
            return hand_level(a.first) < hand_level(b.first);
        }

        return a_type < b_type;
    });

    int64_t total = 0;
    for(int i=0; i < hands.size(); ++i){
        total+=(i+1)*hands[i].second;
    }

    std::cout << "Part1: " << total << std::endl;
}

void part2(std::vector<std::pair<std::string, int64_t>> hands) {

    auto hand_type = [](const std::string& hand) {
        std::map<char, int> counts;
        int joker_count = 0;
        for(char c: hand) {
            if(c=='J') {
                ++joker_count;
                continue;
            }
            ++counts[c];
        }
        std::vector<int> type;
        for(const auto& elem: counts) {
            type.push_back(elem.second);
        }
        std::sort(type.begin(), type.end(), std::greater<int>());
        if(!type.empty()) {
            type[0]+=joker_count;
        } else {
            type.push_back(joker_count);
        }
        return type;
    };

    auto hand_level = [](const std::string& hand) {
        static const std::string order = "J23456789TQKA";
        std::vector<int> result;
        for(char c: hand) {
            auto pos = order.find(c);
            assert(pos!=std::string::npos);
            result.push_back(pos);
        }
        return result;
    };


    std::sort(hands.begin(), hands.end(), [&hand_level, &hand_type](const std::pair<std::string, int64_t>& a, const std::pair<std::string, int64_t>& b) {

        auto a_type = hand_type(a.first);
        auto b_type = hand_type(b.first);

        if(a_type==b_type) {

            auto a_level = hand_level(a.first);
            auto b_level = hand_level(b.first);
            return a_level < b_level;
        }

        return a_type < b_type;
    });

    int64_t total = 0;
    for(int i=0; i < hands.size(); ++i){
        total+=(i+1)*hands[i].second;
    }

    std::cout << "Part2: " << total << std::endl;
}


int main() {

    std::fstream ifs(std::string(SRC_DIR)+"/input.txt");
    std::string line;

    std::vector<std::pair<std::string, int64_t>> hands;
    while(std::getline(ifs, line))
    {
        auto pos = line.find(" ");
        auto hand = line.substr(0, pos);
        int64_t bid = std::stoll(line.substr(pos+1));
        hands.push_back({hand, bid});
    }

    part1(hands);
    part2(hands);
    return 0;
}

File: day07/src/main.rs
use std::collections::{HashMap, HashSet};
use lazy_static::lazy_static;
use regex::{Match, Regex};
use std::fs::File;
use std::io::{self, BufRead};
use std::iter::Map;
use std::path::Path;
use std::slice::Iter;
use std::vec;

fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

fn read() -> Vec<String> {
    let file_path = "input.txt";
    let Ok(lines) = read_lines(file_path) else {
        return vec!();
    };

    return lines.into_iter().map(|x| x.unwrap()).collect();
}

fn hand_type(hand: &str) -> Vec<i64> {
    let mut counts : HashMap<char, i64> = HashMap::new();
    for c in hand.chars() { *counts.entry(c).or_insert(0)+=1; }
    let mut counts = counts.iter().map(|x| *x.1).collect::<Vec<_>>();
    counts.sort_by(|a,b| b.cmp(a));
    counts
}

fn hand_level(hand: &str) -> Vec<usize> {
    let order = "23456789TJQKA";
    let mut result= vec![];
    for c in hand.chars() { result.push(order.find(c).unwrap()); }
    result
}

fn part1(mut hands: Vec<(&str, i64)>) {

    hands.sort_by(|&a, &b| {
        let a_type = hand_type(a.0);
        let b_type = hand_type(b.0);

        if a_type==b_type {
            return hand_level(a.0).cmp(&hand_level(b.0));
        }

        return a_type.cmp(&b_type);
    });

    let total: i64 = hands.iter().enumerate().map(|(idx, &hand)| ((idx+1) as i64) * hand.1).sum();

    println!("part1 {}", total);
}

fn hand_type_part2(hand: &str) -> Vec<i64> {
    let mut counts : HashMap<char, i64> = HashMap::new();
    let mut joker_count = 0i64;
    for c in hand.chars() {
        if c == 'J' {
            joker_count+=1;
            continue
        }
        *counts.entry(c).or_insert(0)+=1;
    }
    let mut counts = counts.iter().map(|x| *x.1).collect::<Vec<_>>();
    counts.sort_by(|a,b| b.cmp(a));
    if counts.is_empty() {
        counts.push(joker_count)
    } else {
        counts[0] += joker_count
    }
    counts
}

fn hand_level_part2(hand: &str) -> Vec<usize> {
    let order = "J23456789TQKA";
    let mut result= vec![];
    for c in hand.chars() { result.push(order.find(c).unwrap()); }
    result
}

fn part2(mut hands: Vec<(&str, i64)>) {

    hands.sort_by(|&a, &b| {
        let a_type = crate::hand_type_part2(a.0);
        let b_type = crate::hand_type_part2(b.0);

        if a_type==b_type {
            return crate::hand_level_part2(a.0).cmp(&crate::hand_level_part2(b.0));
        }

        return a_type.cmp(&b_type);
    });

    let total: i64 = hands.iter().enumerate().map(|(idx, &hand)| ((idx+1) as i64) * hand.1).sum();

    println!("part1 {}", total);
}

fn main() {

    let lines = read();
    let data = lines.iter().map(|line| {
        let parts = line.split(' ').collect::<Vec<_>>();
        (parts[0], parts[1].parse::<i64>().unwrap())
    }).collect::<Vec<_>>();

    part1(data.clone());
    part2(data.clone());
}
